### PR TITLE
Use unbounded channel for event processing

### PR DESCRIPTION
**Pull request by changkun**
### Description:

The `driver.(Canvas).Refresh(fyne.CanvasObject)` and the rendering loop are running on two different threads.
The Refresh will send the input `fyne.CanvasObject` into a buffered channel, then set the canvas to a dirty state so that in the next rendering frame of draw thread the dirty canvas will be repaint on the window as it calls `driver.(*Canvas).FreeDirtyTextures` that drain the refresh queue. All sounds good so far for draw call batching.

However, note that the channel for caching `fyne.CanvasObject` is a buffered channel, which makes the call to `Refresh` may lead to two situations:

1. The channel's buffer is not full, caching is good as always
2. The channel is full, and the requested refresh will be ignored (and hence the draw event may be missed in the select statement, line 264 of internal/driver/common/canvas.go)

Previously, the buffered channel sets an (empirically) large enough size so that it hasn't missed any desired draw calls on the desktop. However, on mobile, especially on iOS, the problem remains suffered and theoretically, the issue may occur also on desktop applications (as application complexity grows).

Instead of keeping grow a "large enough" sized channel in the code, this change introduces an unbounded channel structure that is dedicated to caching `fyne.CanvasObject` between frames.

Lastly, since the data binding also relies on an unbounded channel, the change merges the implementation and introduces an internal/async package that offers generic unbounded channel support.

Updates #950 

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

### Reproducer

```go
package main

import (
	"image/color"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/canvas"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/layout"
	"fyne.io/fyne/v2/widget"
)

func main() {
	app := app.New()
	win := app.NewWindow("Some Window")

	cells := make([]fyne.CanvasObject, 0, 80*25)
	for i := 0; i < 80; i++ {
		for j := 0; j < 25; j++ {
			cells = append(cells, canvas.NewRectangle(color.RGBA{R: 255, G: 255, B: 255, A: 128}))
		}
	}
	container := container.New(layout.NewGridLayout(25), cells...)

	c := 0
	b := widget.NewButton("START", func() {
		for i := 0; i < 80; i++ {
			for j := 0; j < 25; j++ {
				index := i*25 + j
				obj, _ := container.Objects[index].(*canvas.Rectangle)
				if c == 0 {
					obj.FillColor = color.RGBA{R: 20, G: 30, B: 100, A: 128}
				} else if c == 1 {
					obj.FillColor = color.RGBA{R: 100, G: 20, B: 30, A: 128}
				} else if c == 2 {
					obj.FillColor = color.RGBA{R: 30, G: 100, B: 20, A: 128}
				}
				obj.Refresh()
			}
		}
		c = (c + 1) % 3
	})

	win.SetContent(fyne.NewContainerWithLayout(layout.NewBorderLayout(nil, b, nil, nil), container, b))
	win.ShowAndRun()
}
```

Before the fix:

https://user-images.githubusercontent.com/5498964/131047269-f1b89f9c-428a-4c3b-9e72-8855e0523ecd.mp4

After the fix:

https://user-images.githubusercontent.com/5498964/131047282-b48e7ab5-0dd7-445a-8cdf-9b00c7a525b4.mp4

### CODE REVIEW

**Comment by changkun** [comment]
I just rebased the commits and uses the preferred `async` naming. There is a test that failed in the test run, but it seems flaky test that is not relevant to the changes here.

File: internal/chans/gen.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+// Unbounded{{.Name}} is a channel with an unbounded buffer for caching
+// {{.Name}} objects.
+//
+// Delicate dance: One must aware that an unbounded channel may lead
+// to OOM when the consuming speed of the buffer is lower than the
+// producing speed constantly. However, such a channel may be fairly
+// used for event delivering if the consumer of the channel consumes
+// the incoming forever. Especially, rendering and even processing tasks.
+type Unbounded{{.Name}} struct {
```
[the remark this answers is not in the document]
Comment (changkun) [reply internal/chans/gen.go]: Looks like there are no other third-party comments. Hence, use `async.NewXXXChan`.

File: internal/chans/doc.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+// 		},
+// 		"func.go": data{
+// 			Type:    "func()",
+// 			Name:    "Func",
+// 			Imports: "",
+// 		},
+// 	}
+//
+// then run: `go generate ./...` to generate more desired unbounded channels.
+package chans
```
[the remark this answers is not in the document]
Comment (changkun) [reply internal/chans/doc.go]: Resolved in the latest commits

**Comment by andydotxyz** [comment]
Thanks, I re-ran the tests and it's working great now.
For future pull requests can you please avoid force-push once the reviews have started?
We can always squash the result when merging if you feel the history is unhelpful

**Review by andydotxyz — APPROVED** [review]

**Comment by changkun** [comment]
> For future pull requests can you please avoid force-push once the reviews have started?

Sure no problem, I am not entirely sure about the policy in Fyne but now should :) It seems the force push on GitHub does not get properly displayed in contrast to the Gerrit but simply destroys all historical info and no way to see what was changed... (Sigh

**Pull request merged**